### DISCUSSION TITLE
Problems Encountered in ERT Forward

### DISCUSSION THREAD

**Discussion by schaatz** [gimli-org/pyGIMLi#803] [Q&A] [answered]
I want to use ERT to conduct forward modeling of a 3D low-resistivity resistivity model, using the example in this sample.(https://www.pygimli.org/_examples_auto/1_meshing/plot_cad_tutorial.html)

I successfully generated the msh file of the grid, but encountered a problem during the forward modeling process. In the generated dat file with resistivity, all the currents i are 0. 

Could you please tell me what caused this? 

I have attached my forward modeling code and the corresponding files below.
[file.zip](https://github.com/user-attachments/files/18270638/file.zip)

**Comment by halbmy** [comment, marked as answer]
In the forward calculation, a unit current is used and before a recent change to the code, only the apparent resistivity `rhoa` was saved in the data container (unless `calcOnly` is used). This was changed to that both `u` and `i` as well as `r` and `rhoa` are filled.

**Reply by schaatz** [reply]
Thank you for your response!

I checked my Gimli version, which is 1.5.3, and added calcOnly=True as per your reply, changing it to data_mod = ert.simulate(msh_mod, res=rhomap, scheme=scheme, calcOnly=True).

But the generated .dat file still incorrect even not change at all.

**Reply by schaatz** [reply]
Hello, please allow me to restate my problem.
The problem I am currently facing is that when I use the model generated by gmsh to perform ert forward modeling, all the values ​​generated are null values, including current, voltage, apparent resistivity and k. 
The following are my running results and the generated dat file.
```
9
# x y z
6	-20	9.98
6	-15	9.98
6	-10	9.98
6	-5	9.98
6	0	9.98
6	5	9.98
6	10	9.98
6	15	9.98
6	20	9.98
21
# a b m n err i ip iperr k r rhoa u valid 
1	2	3	4	0.00000000000000e+00	0.00000000000000e+00	0.00000000000000e+00	0.00000000000000e+00	nan	0.00000000000000e+00	nan	0.00000000000000e+00	1
1	2	4	5	0.00000000000000e+00	0.00000000000000e+00	0.00000000000000e+00	0.00000000000000e+00	nan	0.00000000000000e+00	nan	0.00000000000000e+00	1
...
```

**Comment by halbmy** [comment]
version v1.5.3 was released November 13, and the commit was done after, on December, 13. So you would have to checkout the source code by git as described on https://www.pygimli.org/installation.html#staying-up-to-date

**Reply by halbmy** [reply]
In the meanwhile, v1.5.4 was released, which allows specifying `ert.simulate(..., current=0.1)` so that both the current and voltage fields are filled. This should finish the original question.  Unfortunately, this was done before adding the noise, so the voltage will be noise-free. I correct this.

**Reply by halbmy** [reply]
I was wrong. This modification was only active for `calcOnly=True`, so I need to do that for the other (default) case as well. So, you still have to set current and voltage after simulate up to now.

**Comment by halbmy** [comment]
Apart from the fact that current and voltage are not set appropriately, I tried a simulation on my computer and it works well:
```
import pygimli as pg
import pygimli.meshtools as mt
from pygimli.physics import ert

scheme = ert.load("BERT_mod.dat")
print(scheme)

mesh = mt.readGmsh("dike_mod.msh2")
print(mesh)
```
yielding
```
Data: Sensors: 152 data: 11175, nonzero entries: ['a', 'b', 'k', 'm', 'n', 'rhoa', 'valid']
Mesh: Nodes: 23727 Cells: 117738 Boundaries: 244457
```

I now attribute resistivities to the three existing regions (`pg.unique(mesh.cellMarkers())` yields 1 2 3)
```
mesh.populate("res", {1:100, 2:100, 3:100})
data = ert.simulate(mesh, scheme=scheme, res=mesh["res"], current=0.1)
data.save("out.data", "a b m n k rhoa")
```
yielding
```
...
14.5	10.49821863	0.391111111
14.5	10.95401125	0.185555556
11175
# a b m n k rhoa
1	2	3	4	-1.42443371918769e+03	1.56596275992202e+03
2	3	4	5	3.97794077394367e+02	3.12693681808745e+02
3	4	5	6	3.98390721989704e+02	2.96577245680928e+02
4	5	6	7	7.45270936365493e+02	5.48345630316307e+02
5	6	7	8	8.29723821609480e+03	6.11052620496144e+03
6	7	8	9	-1.16949814131629e+03	8.53749395482207e+02
7	8	9	10	-6.35050307255084e+02	4.56046823437247e+02
...
```

**Comment by halbmy** [comment]
After fixing the problem by 5c84053, the resulting data file also contains u, i and r (u and r already containing possible noise)
```
14.5	10.95401125	0.185555556
11175
# a b m n u i k rhoa r
1	2	3	4	-1.09935810900000e-01	1.00000000000000e-01	-1.42443371918769e+03	1.56596275992202e+03	-1.09935810900000e+00
2	3	4	5	7.86069224200000e-02	1.00000000000000e-01	3.97794077394367e+02	3.12693681808745e+02	7.86069224200000e-01
```

**Reply by schaatz** [reply]
I found that it couldn't place the electrodes at the boundary points of the model, and when I moved the electrode points, the apparent resistivity values were successfully generated.

Thank you very much for your thoughtful and comprehensive answer! It has been deeply helpful to me. Wishing you the best.
